### PR TITLE
refactor!: extract RPC definitions into here

### PR DESCRIPTION
**Pull request by dignifiedquire**
Add the rpc types and handler to docs, from iroh.

Really, this does add the rpc types and handler to docs, from iroh! Two lines to make check-cc happy.

Breaking changes:

spawn now takes 7 parameters instead of 6

Note:

I reverted the author_default() thing. Not that there was anything wrong with this, but I think breaking less code is good.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -48,6 +48,9 @@ mod ticket;
 #[cfg(feature = "engine")]
 #[cfg_attr(iroh_docsrs, doc(cfg(feature = "engine")))]
 pub mod engine;
+#[cfg(feature = "engine")]
+#[cfg_attr(iroh_docsrs, doc(cfg(feature = "engine")))]
```
Comment (dignifiedquire) [inline src/lib.rs:52]: this should be feature`rpc`

File: src/rpc/client/docs.rs [hunk 2]
```diff
@@ -0,0 +1,868 @@
...
+    /// completed or failed.
+    ///
+    /// It will only be emitted after a [`Self::SyncFinished`] event, never before.
+    ///
+    /// Receiving this event does not guarantee that all content in the document is available. If
+    /// blobs failed to download, this event will still be emitted after all operations completed.
+    PendingContentReady,
+}
+
+impl From<crate::engine::LiveEvent> for LiveEvent {
```
Comment (dignifiedquire) [inline src/rpc/client/docs.rs:569]: I wonder if we can just reuse the same `LiveEvent` type, given we are now in the same crate

File: src/rpc/client/docs.rs [hunk 3]
```diff
@@ -0,0 +1,868 @@
...
+        Err(err) => Err(err.into()),
+    })
+}
+
+#[cfg(test)]
+mod tests {
+    use iroh_blobs::util::local_pool::LocalPool;
+    use iroh_gossip::net::GOSSIP_ALPN;
+    // use rand::RngCore;
+    // use tokio::io::AsyncWriteExt;
```
Comment (dignifiedquire) [inline src/rpc/client/docs.rs:755]: delete?

File: tests/gc.rs [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+    step(&evs).await;
+    assert_eq!(bao_store.entry_status(&h1).await?, EntryStatus::NotFound);
+    assert_eq!(bao_store.entry_status(&h2).await?, EntryStatus::NotFound);
+    assert_eq!(bao_store.entry_status(&hr).await?, EntryStatus::NotFound);
+
+    node.shutdown().await?;
+    Ok(())
+}
+
+mod file {
```
Comment (dignifiedquire) [inline tests/gc.rs:199]: does this need a cfg?

File: src/rpc/client/docs.rs [hunk 5]
```diff
@@ -0,0 +1,868 @@
...
+    use tracing::warn;
+
+    use super::*;
+    use crate::{
+        engine::{DefaultAuthorStorage, Engine},
+        net::DOCS_ALPN,
+    };
+
+    async fn setup_router() -> Result<(
```
Comment (dignifiedquire) [inline src/rpc/client/docs.rs:764]: still needed with the node setup now?

File: tests/util.rs [hunk 6]
```diff
@@ -0,0 +1,401 @@
+#![cfg(feature = "rpc")]
+#![allow(dead_code)]
```
Comment (dignifiedquire) [inline tests/util.rs:2]: is this needed?

**Comment by dignifiedquire** [comment]
some small things, otherwise LGTM

File: tests/util.rs (hunk 6 above)
Comment (rklaehn) [reply tests/util.rs:2]: util is used from the other test rs files. Unfortunately not every test file uses all things in util, so you get dead code even if every single thing is used somewhere. Not sure how to solve this.

File: src/rpc/client/docs.rs [hunk 7]
```diff
@@ -0,0 +1,868 @@
...
+
+impl Stream for ExportFileProgress {
+    type Item = Result<ExportProgress>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        Pin::new(&mut self.stream).poll_next(cx)
+    }
+}
+
+fn flatten<T, E1, E2>(
```
[the remark this answers is not in the document]
Comment (rklaehn) [reply src/rpc/client/docs.rs:736]: I searched for flatten and only found it once.

[the remark this answers is not in the document]
Comment (rklaehn) [reply src/rpc/client/docs.rs:736]: Never mind. Fixed.

File: tests/gc.rs (hunk 4 above)
Comment (rklaehn) [reply tests/gc.rs:199]: Somehow the feature flag for the fs store got lost. Not sure if we need it. The only thing fs-store brings in in terms of deps is redb, and that is very lightweight.

File: src/rpc/client/docs.rs (hunk 5 above)
Comment (rklaehn) [reply src/rpc/client/docs.rs:764]: This entire test can go I think. It only tests that shutdown works correctly, but this is tested several times by the Node being created and dropped.

**Review by rklaehn — APPROVED** [review]
Dig says LGTM

**Pull request merged**